// make_fakes/modules/getChance.js
const { faker } = require('@faker-js/faker');
class GetChance {
    /**
     * let getChance = new GetChance()
     * @date 2022-03-02
     * @returns {any}
     */
    constructor() {

    }
    float() {
        return faker.datatype.float()
    }
    paragraph() {
            return faker.lorem.paragraph()
        }
        /**
         * getChaance.boolean()
         * @date 2022-03-02
         * @returns {any}
         */
    boolean() {
            return faker.datatype.boolean()
        }
        /**
         * getChance.number(1, 100)
         * @date 2022-03-02
         * @param {number} min
         * @param {number} max
         * @returns {number}
         */
    number(min = 1, max = 200) {
        return faker.datatype.number({ min: min, max: max })
    };
    /**
     * getChance.string(10)
     * @date 2022-03-02
     * @param {number} maxLength=10
     * @returns {string}
     */
    string(maxLength = 10) {
        return faker.lorem.word(maxLength)
    };
    /**
     * getChance.email()
     * @date 2022-03-02
     * @returns {string}
     */
    email() {
        return faker.internet.email();
    };
    date() {
            return faker.date.recent()
        }
        /**g
         * getChance.password()
         * @date 2022-03-02
         * @returns {string}
         */
    password() { return faker.internet.password(); };
    /**
     * getChance.past_date()
     * @date 2022-03-02
     * @returns {object}
     */
    past_date() { return faker.date.past(undefined, new Date()) };
    /**
     * getChance.future_date
     * @date 2022-03-02
     * @returns {object}
     */
    future_date() {
        return faker.date.future(
            undefined, new Date()
        );
    };
    /**
     * getChance.number_date
     * @date 2022-03-02
     * @returns {integer}
     */
    number_date() {
        return faker.time.recent()
    };
    /**
     * getChance.phone()
     * @date 2022-03-02
     * @returns {string}
     */
    phone() {
        return faker.phone.phoneNumber()
    };
    /**
     * getChance.getArrayOptions([1, 2, 3])
     * @date 2022-03-02
     * @param {any} array
     * @returns {any}
     */
    getArrayOptions(array) {
        return faker.helpers.randomize(array);
    };
    /**
     * getChance.image
     * @date 2022-03-02
     * @param {any} width=100
     * @param {any} height=100
     * @returns {string}
     */
    image(width = 100, height = 100) {
        return faker.image.lorempicsum.imageUrl(width, height)
    }
}


let getChance = new GetChance()
module.exports = { getChance }
    // let num = getChance.number(1, 100)
    // console.log("🚀 ~ file: playground.js ~ line 62 ~ num", num)
    // let name = getChance.string()
    // console.log("🚀 ~ file: playground.js ~ line 63 ~ name", name)

// const future_date = getChance.future_date()
// console.log(typeof future_date)
// const past_date = getChance.past_date()
// console.log(typeof past_date)
// const number_date = getChance.number_date()
// console.log(typeof number_date)
// const phone = getChance.phone()
// console.log(phone, typeof phone)
// const date = faker.time.recent();
// console.log("🚀 ~ file: playground.js ~ line 58 ~ date", date)
// console.log(new Date(date))
// const arr = getChance.getArrayOptions([1, 2, 3])
// console.log("🚀 ~ file: playground.js ~ line 73 ~ arr", arr)
// console.log(getChance.date())
