// format_date_time_moment/module/getType.js
const { dictionary } = require("./dictionary")
const moment = require("moment")
moment.locale("en")

function countInstances(string, word) {
    return string.split(word).length - 1;
}

function arrStringtoNum(arr) {
    return arr.map(Number);
}

function capitalizeFirstLetter(string) {
    string = string.toLowerCase()
    return string.charAt(0).toUpperCase() + string.slice(1);
}
/**
 * console.log(formatDateTime.secondsToDigital(23423)) 06: 30: 23
 * @author zen-out
 * @date 2022-01-13
 * @param {any} seconds
 * @returns {any}
 */
function secondsToDigital(seconds) {
    var sec_num = parseInt(seconds, 10); // don't forget the second param
    var hours = Math.floor(sec_num / 3600);
    var minutes = Math.floor((sec_num - (hours * 3600)) / 60);
    var seconds = sec_num - (hours * 3600) - (minutes * 60);

    if (hours < 10) {
        hours = "0" + hours;
    }
    if (minutes < 10) {
        minutes = "0" + minutes;
    }
    if (seconds < 10) {
        seconds = "0" + seconds;
    }
    return hours + ':' + minutes + ':' + seconds;
}
/**
* @example
* getType(Date)
*"2022-03-01T19:06:31.047Z"
"00:22:00"
"March 1, 2022"
"Mar 1, 2022"
"2022-03-01"
"01-03-2022"
"2022/03/01"
new Date()
23
* @author zen-out
* @date 2022-01-13
* @param {any} date
* @returns {object} always returns moment 
*/
function getType(date) {
    let finalDate;
    // console.log(date.length)
    if (typeof date === "string") {
        finalDate = date.toUpperCase()
    } else if (typeof date === "number") {
        let seconds = secondsToDigital(date)
        let splitByColon = seconds.split(":")
        let getDigital = arrStringtoNum(splitByColon)
            // console.log(getDigital)
        finalDate = moment({
            hour: parseInt(getDigital[0]),
            minute: parseInt(getDigital[1]),
            second: parseInt(getDigital[2])
        })

        return moment(finalDate)
    } else {
        return moment(date)
    }
    // remove commas
    date = date.replaceAll(",", "")
    let startsWithLetter = new RegExp("^[A-Z]")
    let regular = date.match(startsWithLetter)
    let testDigital = new RegExp(/^([0-1]?\d|2[0-3])(?::([0-5]?\d))?(?::([0-5]?\d))?$/)
        // let testSlash = new RegExp(/[\/]/)
        // var html = date.match(testSlash)
        // console.log(countInstances(date, "/"))
    let slashes = countInstances(date, "/")
    let dashes = countInstances(date, "-")
    var digital = date.match(testDigital);
    let splitted = date.split(" ")
        // let colon = date.
    if (date.length > 20) {
        return moment(date)
    } else if (digital) {
        let splitByColon = date.split(":")
        if (splitByColon.length > 2) {
            let num = arrStringtoNum(splitByColon)
            let time = moment({
                hour: num[0],
                minute: num[1],
                second: num[2]
            })
            return time
        }
    } else if (slashes === 2) {
        // console.log(date)
        let splitted2 = date.split("/")
        console.log("🚀 ~ file: node.js ~ line 145 ~ FormatDateTime ~ getType ~ splitted2", splitted2)
        let slashSplit = arrStringtoNum(splitted2)
        console.log("🚀 ~ file: node.js ~ line 147 ~ FormatDateTime ~ getType ~ slashSplit", slashSplit)
        if (splitted2[0].length > 2) {
            console.log(slashSplit)
            let slashed = moment({
                year: slashSplit[0],
                month: slashSplit[1] - 1,
                day: slashSplit[2]
            })
            return slashed
        } else {
            console.log(slashSplit)
            let slashSplit = moment({
                year: slashSplit[2],
                month: slashSplit[1] - 1,
                day: slashSplit[0]
            })
            return slashSplit
        }
    } else if (dashes === 2) {
        let splitted3 = date.split("-")
        let dashSplit = arrStringtoNum(splitted3)
        if (splitted3[0].length > 2) {
            let dashSplit2 = moment({
                year: dashSplit[0],
                month: dashSplit[1] - 1,
                day: dashSplit[2]
            })
            return dashSplit2
        } else {
            let dashd2 = moment({
                year: dashSplit[2],
                month: dashSplit[1] - 1,
                day: dashSplit[0]
            })
            return dashd2
        }
    } else if (splitted[0].length > 2 && regular) {
        let capitalized = capitalizeFirstLetter(splitted[0])
        splitted[0] = capitalized
        let getMonth = dictionary[splitted[0]]
        let stringed = moment({
            year: parseInt(splitted[2]),
            month: getMonth,
            day: parseInt(splitted[1])
        })
        return stringed;
    }

}

module.exports = { getType }
