// make_legit/index.js
! function(t, e) { "object" == typeof exports && "object" == typeof module ? module.exports = e(require("whats_wrong"), require("when_you_free")) : "function" == typeof define && define.amd ? define(["whats_wrong", "when_you_free"], e) : "object" == typeof exports ? exports.make_legit = e(require("whats_wrong"), require("when_you_free")) : t.make_legit = e(t.whats_wrong, t.when_you_free) }(this, (function(t, e) { return (() => { var n = { 923: function(t, e, n) { var r;! function(o) { "use strict"; var u, i, a, c, f, s, l, p, y, b, h, v, d, g, m, _, j;

                        function w(t) { return null != t }

                        function O(t) { return "number" == typeof t && t > d && t < g }

                        function S(t) { return "number" == typeof t && t % 1 == 0 }

                        function A(t, e) { return O(t) && t > e }

                        function M(t, e) { return O(t) && t < e }

                        function Y(t, e) { return O(t) && t >= e }

                        function T(t, e) { return O(t) && t <= e }

                        function x(t) { return "string" == typeof t }

                        function D(t) { return x(t) && "" !== t }

                        function k(t) { return "[object Object]" === y.call(t) }

                        function E(t, e) { for (var n in t)
                                if (p.call(t, n) && e(n, t[n])) return !0;
                            return !1 }

                        function P(t, e) { try { return t instanceof e } catch (t) { return !1 } }

                        function $(t) { return w(t) && t.length >= 0 }

                        function I(t) { return m ? w(t) && V(t[Symbol.iterator]) : $(t) }

                        function N(t, e) { var n, r; if (!w(t)) return !1; if (j && P(t, Set)) return t.has(e); if (x(t)) return -1 !== t.indexOf(e); if (m && t[Symbol.iterator] && V(t.values)) { n = t.values();
                                do { if ((r = n.next()).value === e) return !0 } while (!r.done); return !1 } return E(t, (function(t, n) { return n === e })) }

                        function F(t, e) { return !!w(t) && (_ && P(t, Map) ? t.has(e) : !(I(t) && !O(+e) || !t[e])) }

                        function V(t) { return "function" == typeof t }

                        function R(t, e) { for (var n in t) p.call(t, n) && e(n, t[n]) }

                        function q(t, e) { var n; for (n = 0; n < t.length; n += 1)
                                if (t[n] === e) return e;
                            return !e }

                        function B(t, e) { var n, r; for (n in t)
                                if (p.call(t, n)) { if (k(r = t[n]) && B(r, e) === e) return e; if (r === e) return e }
                            return !e }

                        function C(t, e) { return R(e, (function(e, n) { t[e] = n })), t }

                        function L(t, e) { return function() { var n = arguments,
                                    r = t.l || t.length,
                                    o = n[r],
                                    u = n[r + 1]; return U(t.apply(null, n), D(o) ? o : e.replace("{a}", z(n[0])).replace("{e}", z(n[1])).replace("{e2}", z(n[2])).replace("{t}", (function() { var t = n[1]; return t && t.name ? t.name : t })), V(u) ? u : TypeError), n[0] } }

                        function z(t) { return function() { return x(t) ? '"' + t.replace(/\\/g, "\\\\").replace(/"/g, '\\"') + '"' : t && !0 !== t && t.constructor && !P(t, RegExp) && "number" != typeof t ? t.constructor.name : t } }

                        function U(t, e, n) { if (t) return t; throw new(n || Error)(e || "assert failed") }

                        function J(t) { var e = function() { return W(t.apply(null, arguments)) }; return e.l = t.length, e }

                        function W(t) { return !t }

                        function G(t, e, n) { var r = function() { var r, o; if (r = arguments[0], "maybe" === t && f.assigned(r)) return !0; if (!e(r)) return !1;
                                r = K(e, r), o = h.call(arguments, 1); try { r.forEach((function(e) { if (("maybe" !== t || w(e)) && !n.apply(null, [e].concat(o))) throw 0 })) } catch (t) { return !1 } return !0 }; return r.l = n.length, r }

                        function K(t, e) { switch (t) {
                                case $:
                                    return h.call(e);
                                case k:
                                    return b(e).map((function(t) { return e[t] }));
                                default:
                                    return e } }

                        function H(t, e) { return Z([t, i, e, ""]) }

                        function Z(t) { var e, n, r; return e = t.shift(), n = t.pop(), r = t.pop(), R(t.pop(), (function(o, i) { var a = u[o];
                                a && n && (a = a.replace("to", n + "to")), Object.defineProperty(r, o, { configurable: !1, enumerable: !0, writable: !1, value: e.apply(null, t.concat(i, a)) }) })), r }

                        function Q(t, e, n) { return Z([t, e, {}, n]) }

                        function X(t, e) { l.forEach((function(n) { t[n].of = Q(e, i[n].of) })) }
                        u = {}, i = {}, [{ n: "equal", f: function(t, e) { return t === e }, s: "equal {e}" }, { n: "undefined", f: function(t) { return void 0 === t }, s: "be undefined" }, { n: "null", f: function(t) { return null === t }, s: "be null" }, { n: "assigned", f: w, s: "be assigned" }, { n: "primitive", f: function(t) { var e; switch (t) {
                                        case null:
                                        case void 0:
                                        case !1:
                                        case !0:
                                            return !0 } return "string" == (e = typeof t) || "number" === e || m && "symbol" === e }, s: "be primitive type" }, { n: "contains", f: N, s: "contain {e}" }, { n: "in", f: function(t, e) { return N(e, t) }, s: "be in {e}" }, { n: "containsKey", f: F, s: "contain key {e}" }, { n: "keyIn", f: function(t, e) { return F(e, t) }, s: "be key in {e}" }, { n: "zero", f: function(t) { return 0 === t }, s: "be 0" }, { n: "one", f: function(t) { return 1 === t }, s: "be 1" }, { n: "infinity", f: function(t) { return t === d || t === g }, s: "be infinity" }, { n: "number", f: O, s: "be Number" }, { n: "integer", f: S, s: "be integer" }, { n: "float", f: function(t) { return O(t) && t % 1 != 0 }, s: "be non-integer number" }, { n: "even", f: function(t) { return "number" == typeof t && t % 2 == 0 }, s: "be even number" }, { n: "odd", f: function(t) { return S(t) && t % 2 != 0 }, s: "be odd number" }, { n: "greater", f: A, s: "be greater than {e}" }, { n: "less", f: M, s: "be less than {e}" }, { n: "between", f: function(t, e, n) { return e < n ? A(t, e) && t < n : M(t, e) && t > n }, s: "be between {e} and {e2}" }, { n: "greaterOrEqual", f: Y, s: "be greater than or equal to {e}" }, { n: "lessOrEqual", f: T, s: "be less than or equal to {e}" }, { n: "inRange", f: function(t, e, n) { return e < n ? Y(t, e) && t <= n : T(t, e) && t >= n }, s: "be in the range {e} to {e2}" }, { n: "positive", f: function(t) { return A(t, 0) }, s: "be positive number" }, { n: "negative", f: function(t) { return M(t, 0) }, s: "be negative number" }, { n: "string", f: x, s: "be String" }, { n: "emptyString", f: function(t) { return "" === t }, s: "be empty string" }, { n: "nonEmptyString", f: D, s: "be non-empty string" }, { n: "match", f: function(t, e) { return x(t) && !!t.match(e) }, s: "match {e}" }, { n: "boolean", f: function(t) { return !1 === t || !0 === t }, s: "be Boolean" }, { n: "object", f: k, s: "be Object" }, { n: "emptyObject", f: function(t) { return k(t) && !E(t, (function() { return !0 })) }, s: "be empty object" }, { n: "nonEmptyObject", f: function(t) { return k(t) && E(t, (function() { return !0 })) }, s: "be non-empty object" }, { n: "instanceStrict", f: P, s: "be instanceof {t}" }, { n: "thenable", f: function(t) { return w(t) && V(t.then) }, s: "be promise-like" }, { n: "instance", f: function(t, e) { try { return P(t, e) || t.constructor.name === e.name || y.call(t) === "[object " + e.name + "]" } catch (t) { return !1 } }, s: "be {t}" }, { n: "like", f: function t(e, n) { var r; for (r in n)
                                        if (p.call(n, r)) { if (!1 === p.call(e, r) || typeof e[r] != typeof n[r]) return !1; if (k(e[r]) && !1 === t(e[r], n[r])) return !1 }
                                    return !0 }, s: "be like {e}" }, { n: "array", f: function(t) { return v(t) }, s: "be Array" }, { n: "emptyArray", f: function(t) { return v(t) && 0 === t.length }, s: "be empty array" }, { n: "nonEmptyArray", f: function(t) { return v(t) && t.length > 0 }, s: "be non-empty array" }, { n: "arrayLike", f: $, s: "be array-like" }, { n: "iterable", f: I, s: "be iterable" }, { n: "date", f: function(t) { return P(t, Date) && S(t.getTime()) }, s: "be valid Date" }, { n: "function", f: V, s: "be Function" }, { n: "hasLength", f: function(t, e) { return w(t) && t.length === e }, s: "have length {e}" }, { n: "throws", f: function(t) { if (!V(t)) return !1; try { t() } catch (t) { return !0 } return !1 }, s: "throw" }].map((function(t) { var e = t.n;
                                u[e] = "assert failed: expected {a} to " + t.s, i[e] = t.f })), a = { map: function t(e, n) { var r; if (r = v(e) ? [] : {}, V(n)) R(e, (function(t, e) { r[t] = n(e) }));
                                    else { v(n) || c.object(n); var o = b(e || {});
                                        R(n, (function(n, u) { o.some((function(t, e) { return t === n && (o.splice(e, 1), !0) })), V(u) ? f.assigned(e) ? r[n] = !!u.m : r[n] = u(e[n]) : r[n] = t(e[n], u) })) } return r }, all: function(t) { return v(t) ? q(t, !1) : (c.object(t), B(t, !1)) }, any: function(t) { return v(t) ? q(t, !0) : (c.object(t), B(t, !0)) } }, l = ["array", "arrayLike", "iterable", "object"], p = Object.prototype.hasOwnProperty, y = Object.prototype.toString, b = Object.keys, h = Array.prototype.slice, v = Array.isArray, d = Number.NEGATIVE_INFINITY, g = Number.POSITIVE_INFINITY, m = "function" == typeof Symbol, _ = "function" == typeof Map, j = "function" == typeof Set, a = C(a, i), c = H(L, U), f = H(J, W), s = H((function(t) { var e = function() { return !!f.assigned(arguments[0]) || t.apply(null, arguments) }; return e.l = t.length, e.m = !0, e }), (function(t) { return !1 === w(t) || t })), c.not = Q(L, f, "not "), c.maybe = Q(L, s, "maybe "), l.forEach((function(t) { i[t].of = Z([G.bind(null, null), i[t], i, {}, ""]) })), X(c, L), X(f, J), l.forEach((function(t) { s[t].of = Z([G.bind(null, "maybe"), i[t], i, {}, ""]), c.maybe[t].of = Q(L, s[t].of), c.not[t].of = Q(L, f[t].of) })),
                            function(o) { void 0 === (r = function() { return o }.call(e, n, e, t)) || (t.exports = r) }(C(a, { assert: c, not: f, maybe: s })) }() }, 826: (t, e, n) => { const r = n(245),
                        o = n(146),
                        u = n(684);
                    t.exports = function t(e, { cleanKeys: n = [], cleanValues: i = [], emptyArrays: a = !0, emptyObjects: c = !0, emptyStrings: f = !0, NaNValues: s = !1, nullValues: l = !0, undefinedValues: p = !0 } = {}) { return u(e, ((e, u, y) => { if (!n.includes(y) && ((Array.isArray(u) || o(u)) && (u = t(u, { NaNValues: s, cleanKeys: n, cleanValues: i, emptyArrays: a, emptyObjects: c, emptyStrings: f, nullValues: l, undefinedValues: p })), !i.includes(u) && !(c && o(u) && r(u)) && (!a || !Array.isArray(u) || u.length) && !(f && "" === u || s && Number.isNaN(u) || l && null === u || p && void 0 === u))) return Array.isArray(e) ? e.push(u) : void(e[y] = u) })) } }, 53: function(t) { t.exports = (() => { "use strict"; var t = { d: (e, n) => { for (var r in n) t.o(n, r) && !t.o(e, r) && Object.defineProperty(e, r, { enumerable: !0, get: n[r] }) }, o: (t, e) => Object.prototype.hasOwnProperty.call(t, e), r: t => { "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(t, Symbol.toStringTag, { value: "Module" }), Object.defineProperty(t, "__esModule", { value: !0 }) } },
                            e = {};
                        t.r(e), t.d(e, { containsThis: () => y, dashDDMMYYYY: () => s, dashYYYYMMDD: () => f, git: () => r, hhmm: () => o, hhmmss: () => u, hyphenDDMMYYYY: () => c, hyphenYYYYMMDD: () => a, knexDate: () => p, matchWords: () => b, node_modules: () => n, stringDate: () => l, yyyymmdd: () => i }); var n = /^(?:.*[\\\/])?node_modules(?:[\\\/].*)?$/,
                            r = /^(?:.*[\\\/])?.git(?:[\\\/].*)?$/,
                            o = /^(2[0-3]|[01][0-9]):?([0-5][0-9])$/,
                            u = /^(2[0-3]|[01][0-9]):?([0-5][0-9]):?([0-5][0-9])$/,
                            i = /((19|20)\d{2}(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01]))/,
                            a = /\d{4}\-(0?[1-9]|1[012])\-(0?[1-9]|[12][0-9]|3[01])*/,
                            c = /^\d{1,2}\-\d{1,2}\-\d{4}$/,
                            f = /^\d{4}\/\d{1,2}\/\d{1,2}$/,
                            s = /^\d{1,2}\/\d{1,2}\/\d{4}$/,
                            l = /(Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\s+(\d{1,2})\s+(\d{4})/,
                            p = /\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[1-2]\d|3[0-1])T(?:[0-1]\d|2[0-3]):[0-5]\d:[0-5]\dZ/;

                        function y(t, e) { return function(t) { var e = new RegExp(/^[A-Z]/),
                                    n = new RegExp(/^[a-z]/),
                                    r = e.test(t),
                                    o = n.test(t); return r && o }(t) && (t = function(t) { return t.charAt(0).toUpperCase() + t.slice(1) }(t)), new RegExp(e).test(t) }

                        function b(t, e) { for (var n = /[(){[*+?.\\^$|]/g, r = 0; r < e.length; r++) e[r] = e[r].replace(n, "\\$&"); var o = new RegExp("\\b(?:" + e.join("|") + ")\\b", "gi"); return t.match(o) || [] } return e })() }, 245: (t, e, n) => { t = n.nmd(t); var r, o, u, i = "[object Map]",
                        a = "[object Promise]",
                        c = "[object Set]",
                        f = "[object WeakMap]",
                        s = "[object DataView]",
                        l = /^\[object .+?Constructor\]$/,
                        p = "object" == typeof n.g && n.g && n.g.Object === Object && n.g,
                        y = "object" == typeof self && self && self.Object === Object && self,
                        b = p || y || Function("return this")(),
                        h = e && !e.nodeType && e,
                        v = h && t && !t.nodeType && t,
                        d = v && v.exports === h,
                        g = Function.prototype,
                        m = Object.prototype,
                        _ = b["__core-js_shared__"],
                        j = (r = /[^.]+$/.exec(_ && _.keys && _.keys.IE_PROTO || "")) ? "Symbol(src)_1." + r : "",
                        w = g.toString,
                        O = m.hasOwnProperty,
                        S = m.toString,
                        A = RegExp("^" + w.call(O).replace(/[\\^$.*+?()[\]{}|]/g, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"),
                        M = d ? b.Buffer : void 0,
                        Y = m.propertyIsEnumerable,
                        T = M ? M.isBuffer : void 0,
                        x = (o = Object.keys, u = Object, function(t) { return o(u(t)) }),
                        D = B(b, "DataView"),
                        k = B(b, "Map"),
                        E = B(b, "Promise"),
                        P = B(b, "Set"),
                        $ = B(b, "WeakMap"),
                        I = !Y.call({ valueOf: 1 }, "valueOf"),
                        N = L(D),
                        F = L(k),
                        V = L(E),
                        R = L(P),
                        q = L($);

                    function B(t, e) { var n = function(t, e) { return null == t ? void 0 : t[e] }(t, e); return function(t) { if (!G(t) || function(t) { return !!j && j in t }(t)) return !1; var e = W(t) || function(t) { var e = !1; if (null != t && "function" != typeof t.toString) try { e = !!(t + "") } catch (t) {}
                                return e }(t) ? A : l; return e.test(L(t)) }(n) ? n : void 0 } var C = function(t) { return S.call(t) };

                    function L(t) { if (null != t) { try { return w.call(t) } catch (t) {} try { return t + "" } catch (t) {} } return "" }(D && C(new D(new ArrayBuffer(1))) != s || k && C(new k) != i || E && C(E.resolve()) != a || P && C(new P) != c || $ && C(new $) != f) && (C = function(t) { var e = S.call(t),
                            n = "[object Object]" == e ? t.constructor : void 0,
                            r = n ? L(n) : void 0; if (r) switch (r) {
                            case N:
                                return s;
                            case F:
                                return i;
                            case V:
                                return a;
                            case R:
                                return c;
                            case q:
                                return f }
                        return e }); var z = Array.isArray;

                    function U(t) { return null != t && function(t) { return "number" == typeof t && t > -1 && t % 1 == 0 && t <= 9007199254740991 }(t.length) && !W(t) } var J = T || function() { return !1 };

                    function W(t) { var e = G(t) ? S.call(t) : ""; return "[object Function]" == e || "[object GeneratorFunction]" == e }

                    function G(t) { var e = typeof t; return !!t && ("object" == e || "function" == e) }
                    t.exports = function(t) { if (U(t) && (z(t) || "string" == typeof t || "function" == typeof t.splice || J(t) || function(t) { return function(t) { return function(t) { return !!t && "object" == typeof t }(t) && U(t) }(t) && O.call(t, "callee") && (!Y.call(t, "callee") || "[object Arguments]" == S.call(t)) }(t))) return !t.length; var e = C(t); if (e == i || e == c) return !t.size; if (I || function(t) { var e = t && t.constructor; return t === ("function" == typeof e && e.prototype || m) }(t)) return !x(t).length; for (var n in t)
                            if (O.call(t, n)) return !1;
                        return !0 } }, 146: t => { var e, n, r = Function.prototype,
                        o = Object.prototype,
                        u = r.toString,
                        i = o.hasOwnProperty,
                        a = u.call(Object),
                        c = o.toString,
                        f = (e = Object.getPrototypeOf, n = Object, function(t) { return e(n(t)) });
                    t.exports = function(t) { if (! function(t) { return !!t && "object" == typeof t }(t) || "[object Object]" != c.call(t) || function(t) { var e = !1; if (null != t && "function" != typeof t.toString) try { e = !!(t + "") } catch (t) {}
                                return e }(t)) return !1; var e = f(t); if (null === e) return !0; var n = i.call(e, "constructor") && e.constructor; return "function" == typeof n && n instanceof n && u.call(n) == a } }, 684: (t, e, n) => { t = n.nmd(t); var r = "__lodash_hash_undefined__",
                        o = 9007199254740991,
                        u = "[object Arguments]",
                        i = "[object Array]",
                        a = "[object Boolean]",
                        c = "[object Date]",
                        f = "[object Error]",
                        s = "[object Function]",
                        l = "[object Map]",
                        p = "[object Number]",
                        y = "[object Object]",
                        b = "[object Promise]",
                        h = "[object RegExp]",
                        v = "[object Set]",
                        d = "[object String]",
                        g = "[object Symbol]",
                        m = "[object WeakMap]",
                        _ = "[object ArrayBuffer]",
                        j = "[object DataView]",
                        w = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/,
                        O = /^\w*$/,
                        S = /^\./,
                        A = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g,
                        M = /\\(\\)?/g,
                        Y = /^\[object .+?Constructor\]$/,
                        T = /^(?:0|[1-9]\d*)$/,
                        x = {};
                    x["[object Float32Array]"] = x["[object Float64Array]"] = x["[object Int8Array]"] = x["[object Int16Array]"] = x["[object Int32Array]"] = x["[object Uint8Array]"] = x["[object Uint8ClampedArray]"] = x["[object Uint16Array]"] = x["[object Uint32Array]"] = !0, x[u] = x[i] = x[_] = x[a] = x[j] = x[c] = x[f] = x[s] = x[l] = x[p] = x[y] = x[h] = x[v] = x[d] = x[m] = !1; var D = "object" == typeof n.g && n.g && n.g.Object === Object && n.g,
                        k = "object" == typeof self && self && self.Object === Object && self,
                        E = D || k || Function("return this")(),
                        P = e && !e.nodeType && e,
                        $ = P && t && !t.nodeType && t,
                        I = $ && $.exports === P && D.process,
                        N = function() { try { return I && I.binding("util") } catch (t) {} }(),
                        F = N && N.isTypedArray;

                    function V(t, e) { for (var n = -1, r = t ? t.length : 0; ++n < r && !1 !== e(t[n], n, t);); return t }

                    function R(t, e) { for (var n = -1, r = t ? t.length : 0; ++n < r;)
                            if (e(t[n], n, t)) return !0;
                        return !1 }

                    function q(t) { var e = !1; if (null != t && "function" != typeof t.toString) try { e = !!(t + "") } catch (t) {}
                        return e }

                    function B(t) { var e = -1,
                            n = Array(t.size); return t.forEach((function(t, r) { n[++e] = [r, t] })), n }

                    function C(t, e) { return function(n) { return t(e(n)) } }

                    function L(t) { var e = -1,
                            n = Array(t.size); return t.forEach((function(t) { n[++e] = t })), n } var z, U = Array.prototype,
                        J = Function.prototype,
                        W = Object.prototype,
                        G = E["__core-js_shared__"],
                        K = (z = /[^.]+$/.exec(G && G.keys && G.keys.IE_PROTO || "")) ? "Symbol(src)_1." + z : "",
                        H = J.toString,
                        Z = W.hasOwnProperty,
                        Q = W.toString,
                        X = RegExp("^" + H.call(Z).replace(/[\\^$.*+?()[\]{}|]/g, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"),
                        tt = E.Symbol,
                        et = E.Uint8Array,
                        nt = C(Object.getPrototypeOf, Object),
                        rt = Object.create,
                        ot = W.propertyIsEnumerable,
                        ut = U.splice,
                        it = C(Object.keys, Object),
                        at = $t(E, "DataView"),
                        ct = $t(E, "Map"),
                        ft = $t(E, "Promise"),
                        st = $t(E, "Set"),
                        lt = $t(E, "WeakMap"),
                        pt = $t(Object, "create"),
                        yt = Ct(at),
                        bt = Ct(ct),
                        ht = Ct(ft),
                        vt = Ct(st),
                        dt = Ct(lt),
                        gt = tt ? tt.prototype : void 0,
                        mt = gt ? gt.valueOf : void 0,
                        _t = gt ? gt.toString : void 0;

                    function jt(t) { var e = -1,
                            n = t ? t.length : 0; for (this.clear(); ++e < n;) { var r = t[e];
                            this.set(r[0], r[1]) } }

                    function wt(t) { var e = -1,
                            n = t ? t.length : 0; for (this.clear(); ++e < n;) { var r = t[e];
                            this.set(r[0], r[1]) } }

                    function Ot(t) { var e = -1,
                            n = t ? t.length : 0; for (this.clear(); ++e < n;) { var r = t[e];
                            this.set(r[0], r[1]) } }

                    function St(t) { var e = -1,
                            n = t ? t.length : 0; for (this.__data__ = new Ot; ++e < n;) this.add(t[e]) }

                    function At(t) { this.__data__ = new wt(t) }

                    function Mt(t, e) { for (var n = t.length; n--;)
                            if (zt(t[n][0], e)) return n;
                        return -1 }
                    jt.prototype.clear = function() { this.__data__ = pt ? pt(null) : {} }, jt.prototype.delete = function(t) { return this.has(t) && delete this.__data__[t] }, jt.prototype.get = function(t) { var e = this.__data__; if (pt) { var n = e[t]; return n === r ? void 0 : n } return Z.call(e, t) ? e[t] : void 0 }, jt.prototype.has = function(t) { var e = this.__data__; return pt ? void 0 !== e[t] : Z.call(e, t) }, jt.prototype.set = function(t, e) { return this.__data__[t] = pt && void 0 === e ? r : e, this }, wt.prototype.clear = function() { this.__data__ = [] }, wt.prototype.delete = function(t) { var e = this.__data__,
                            n = Mt(e, t); return !(n < 0 || (n == e.length - 1 ? e.pop() : ut.call(e, n, 1), 0)) }, wt.prototype.get = function(t) { var e = this.__data__,
                            n = Mt(e, t); return n < 0 ? void 0 : e[n][1] }, wt.prototype.has = function(t) { return Mt(this.__data__, t) > -1 }, wt.prototype.set = function(t, e) { var n = this.__data__,
                            r = Mt(n, t); return r < 0 ? n.push([t, e]) : n[r][1] = e, this }, Ot.prototype.clear = function() { this.__data__ = { hash: new jt, map: new(ct || wt), string: new jt } }, Ot.prototype.delete = function(t) { return Pt(this, t).delete(t) }, Ot.prototype.get = function(t) { return Pt(this, t).get(t) }, Ot.prototype.has = function(t) { return Pt(this, t).has(t) }, Ot.prototype.set = function(t, e) { return Pt(this, t).set(t, e), this }, St.prototype.add = St.prototype.push = function(t) { return this.__data__.set(t, r), this }, St.prototype.has = function(t) { return this.__data__.has(t) }, At.prototype.clear = function() { this.__data__ = new wt }, At.prototype.delete = function(t) { return this.__data__.delete(t) }, At.prototype.get = function(t) { return this.__data__.get(t) }, At.prototype.has = function(t) { return this.__data__.has(t) }, At.prototype.set = function(t, e) { var n = this.__data__; if (n instanceof wt) { var r = n.__data__; if (!ct || r.length < 199) return r.push([t, e]), this;
                            n = this.__data__ = new Ot(r) } return n.set(t, e), this };

                    function Yt(t, e) { return t && function(t, e, n) { for (var r = -1, o = Object(t), u = n(t), i = u.length; i--;) { var a = u[++r]; if (!1 === e(o[a], a, o)) break } return t }(t, e, ee) }

                    function Tt(t, e) { for (var n = 0, r = (e = Ft(e, t) ? [e] : kt(e)).length; null != t && n < r;) t = t[Bt(e[n++])]; return n && n == r ? t : void 0 }

                    function xt(t, e) { return null != t && e in Object(t) }

                    function Dt(t, e, n, r, o) { return t === e || (null == t || null == e || !Ht(t) && !Zt(e) ? t != t && e != e : function(t, e, n, r, o, s) { var b = Jt(t),
                                m = Jt(e),
                                w = i,
                                O = i;
                            b || (w = (w = It(t)) == u ? y : w), m || (O = (O = It(e)) == u ? y : O); var S = w == y && !q(t),
                                A = O == y && !q(e),
                                M = w == O; if (M && !S) return s || (s = new At), b || te(t) ? Et(t, e, n, r, o, s) : function(t, e, n, r, o, u, i) { switch (n) {
                                    case j:
                                        if (t.byteLength != e.byteLength || t.byteOffset != e.byteOffset) return !1;
                                        t = t.buffer, e = e.buffer;
                                    case _:
                                        return !(t.byteLength != e.byteLength || !r(new et(t), new et(e)));
                                    case a:
                                    case c:
                                    case p:
                                        return zt(+t, +e);
                                    case f:
                                        return t.name == e.name && t.message == e.message;
                                    case h:
                                    case d:
                                        return t == e + "";
                                    case l:
                                        var s = B;
                                    case v:
                                        var y = 2 & u; if (s || (s = L), t.size != e.size && !y) return !1; var b = i.get(t); if (b) return b == e;
                                        u |= 1, i.set(t, e); var m = Et(s(t), s(e), r, o, u, i); return i.delete(t), m;
                                    case g:
                                        if (mt) return mt.call(t) == mt.call(e) } return !1 }(t, e, w, n, r, o, s); if (!(2 & o)) { var Y = S && Z.call(t, "__wrapped__"),
                                    T = A && Z.call(e, "__wrapped__"); if (Y || T) { var x = Y ? t.value() : t,
                                        D = T ? e.value() : e; return s || (s = new At), n(x, D, r, o, s) } } return !!M && (s || (s = new At), function(t, e, n, r, o, u) { var i = 2 & o,
                                    a = ee(t),
                                    c = a.length; if (c != ee(e).length && !i) return !1; for (var f = c; f--;) { var s = a[f]; if (!(i ? s in e : Z.call(e, s))) return !1 } var l = u.get(t); if (l && u.get(e)) return l == e; var p = !0;
                                u.set(t, e), u.set(e, t); for (var y = i; ++f < c;) { var b = t[s = a[f]],
                                        h = e[s]; if (r) var v = i ? r(h, b, s, e, t, u) : r(b, h, s, t, e, u); if (!(void 0 === v ? b === h || n(b, h, r, o, u) : v)) { p = !1; break }
                                    y || (y = "constructor" == s) } if (p && !y) { var d = t.constructor,
                                        g = e.constructor;
                                    d == g || !("constructor" in t) || !("constructor" in e) || "function" == typeof d && d instanceof d && "function" == typeof g && g instanceof g || (p = !1) } return u.delete(t), u.delete(e), p }(t, e, n, r, o, s)) }(t, e, Dt, n, r, o)) }

                    function kt(t) { return Jt(t) ? t : qt(t) }

                    function Et(t, e, n, r, o, u) { var i = 2 & o,
                            a = t.length,
                            c = e.length; if (a != c && !(i && c > a)) return !1; var f = u.get(t); if (f && u.get(e)) return f == e; var s = -1,
                            l = !0,
                            p = 1 & o ? new St : void 0; for (u.set(t, e), u.set(e, t); ++s < a;) { var y = t[s],
                                b = e[s]; if (r) var h = i ? r(b, y, s, e, t, u) : r(y, b, s, t, e, u); if (void 0 !== h) { if (h) continue;
                                l = !1; break } if (p) { if (!R(e, (function(t, e) { if (!p.has(e) && (y === t || n(y, t, r, o, u))) return p.add(e) }))) { l = !1; break } } else if (y !== b && !n(y, b, r, o, u)) { l = !1; break } } return u.delete(t), u.delete(e), l }

                    function Pt(t, e) { var n, r, o = t.__data__; return ("string" == (r = typeof(n = e)) || "number" == r || "symbol" == r || "boolean" == r ? "__proto__" !== n : null === n) ? o["string" == typeof e ? "string" : "hash"] : o.map }

                    function $t(t, e) { var n = function(t, e) { return null == t ? void 0 : t[e] }(t, e); return function(t) { return !(!Ht(t) || (e = t, K && K in e)) && (Gt(t) || q(t) ? X : Y).test(Ct(t)); var e }(n) ? n : void 0 } var It = function(t) { return Q.call(t) };

                    function Nt(t, e) { return !!(e = null == e ? o : e) && ("number" == typeof t || T.test(t)) && t > -1 && t % 1 == 0 && t < e }

                    function Ft(t, e) { if (Jt(t)) return !1; var n = typeof t; return !("number" != n && "symbol" != n && "boolean" != n && null != t && !Qt(t)) || O.test(t) || !w.test(t) || null != e && t in Object(e) }

                    function Vt(t) { return t == t && !Ht(t) }

                    function Rt(t, e) { return function(n) { return null != n && n[t] === e && (void 0 !== e || t in Object(n)) } }(at && It(new at(new ArrayBuffer(1))) != j || ct && It(new ct) != l || ft && It(ft.resolve()) != b || st && It(new st) != v || lt && It(new lt) != m) && (It = function(t) { var e = Q.call(t),
                            n = e == y ? t.constructor : void 0,
                            r = n ? Ct(n) : void 0; if (r) switch (r) {
                            case yt:
                                return j;
                            case bt:
                                return l;
                            case ht:
                                return b;
                            case vt:
                                return v;
                            case dt:
                                return m }
                        return e }); var qt = Lt((function(t) { var e;
                        t = null == (e = t) ? "" : function(t) { if ("string" == typeof t) return t; if (Qt(t)) return _t ? _t.call(t) : ""; var e = t + ""; return "0" == e && 1 / t == -1 / 0 ? "-0" : e }(e); var n = []; return S.test(t) && n.push(""), t.replace(A, (function(t, e, r, o) { n.push(r ? o.replace(M, "$1") : e || t) })), n }));

                    function Bt(t) { if ("string" == typeof t || Qt(t)) return t; var e = t + ""; return "0" == e && 1 / t == -1 / 0 ? "-0" : e }

                    function Ct(t) { if (null != t) { try { return H.call(t) } catch (t) {} try { return t + "" } catch (t) {} } return "" }

                    function Lt(t, e) { if ("function" != typeof t || e && "function" != typeof e) throw new TypeError("Expected a function"); var n = function() { var r = arguments,
                                o = e ? e.apply(this, r) : r[0],
                                u = n.cache; if (u.has(o)) return u.get(o); var i = t.apply(this, r); return n.cache = u.set(o, i), i }; return n.cache = new(Lt.Cache || Ot), n }

                    function zt(t, e) { return t === e || t != t && e != e }

                    function Ut(t) { return function(t) { return Zt(t) && Wt(t) }(t) && Z.call(t, "callee") && (!ot.call(t, "callee") || Q.call(t) == u) }
                    Lt.Cache = Ot; var Jt = Array.isArray;

                    function Wt(t) { return null != t && Kt(t.length) && !Gt(t) }

                    function Gt(t) { var e = Ht(t) ? Q.call(t) : ""; return e == s || "[object GeneratorFunction]" == e }

                    function Kt(t) { return "number" == typeof t && t > -1 && t % 1 == 0 && t <= o }

                    function Ht(t) { var e = typeof t; return !!t && ("object" == e || "function" == e) }

                    function Zt(t) { return !!t && "object" == typeof t }

                    function Qt(t) { return "symbol" == typeof t || Zt(t) && Q.call(t) == g } var Xt, te = F ? (Xt = F, function(t) { return Xt(t) }) : function(t) { return Zt(t) && Kt(t.length) && !!x[Q.call(t)] };

                    function ee(t) { return Wt(t) ? function(t, e) { var n = Jt(t) || Ut(t) ? function(t, e) { for (var n = -1, r = Array(t); ++n < t;) r[n] = e(n); return r }(t.length, String) : [],
                                r = n.length,
                                o = !!r; for (var u in t) !e && !Z.call(t, u) || o && ("length" == u || Nt(u, r)) || n.push(u); return n }(t) : function(t) { if (n = (e = t) && e.constructor, e !== ("function" == typeof n && n.prototype || W)) return it(t); var e, n, r = []; for (var o in Object(t)) Z.call(t, o) && "constructor" != o && r.push(o); return r }(t) }

                    function ne(t) { return t }
                    t.exports = function(t, e, n) { var r, o, u, i, a, c, f = Jt(t) || te(t); if (e = "function" == typeof(o = e) ? o : null == o ? ne : "object" == typeof o ? Jt(o) ? function(t, e) { return Ft(t) && Vt(e) ? Rt(Bt(t), e) : function(n) { var r = function(t, e, n) { var r = null == t ? void 0 : Tt(t, e); return void 0 === r ? void 0 : r }(n, t); return void 0 === r && r === e ? function(t, e) { return null != t && function(t, e, n) { for (var r, o = -1, u = (e = Ft(e, t) ? [e] : kt(e)).length; ++o < u;) { var i = Bt(e[o]); if (!(r = null != t && n(t, i))) break;
                                                t = t[i] } return r || !!(u = t ? t.length : 0) && Kt(u) && Nt(i, u) && (Jt(t) || Ut(t)) }(t, e, xt) }(n, t) : Dt(e, r, void 0, 3) } }(o[0], o[1]) : (i = function(t) { for (var e = ee(t), n = e.length; n--;) { var r = e[n],
                                        o = t[r];
                                    e[n] = [r, o, Vt(o)] } return e }(u = o), 1 == i.length && i[0][2] ? Rt(i[0][0], i[0][1]) : function(t) { return t === u || function(t, e, n, r) { var o = n.length,
                                        u = o; if (null == t) return !u; for (t = Object(t); o--;) { var i = n[o]; if (i[2] ? i[1] !== t[i[0]] : !(i[0] in t)) return !1 } for (; ++o < u;) { var a = (i = n[o])[0],
                                            c = t[a],
                                            f = i[1]; if (i[2]) { if (void 0 === c && !(a in t)) return !1 } else { var s, l = new At; if (!(void 0 === s ? Dt(f, c, r, 3, l) : s)) return !1 } } return !0 }(t, 0, i) }) : Ft(a = o) ? (c = Bt(a), function(t) { return null == t ? void 0 : t[c] }) : function(t) { return function(e) { return Tt(e, t) } }(a), null == n)
                            if (f || Ht(t)) { var s = t.constructor;
                                n = f ? Jt(t) ? new s : [] : Gt(s) && Ht(r = nt(t)) ? rt(r) : {} } else n = {};
                        return (f ? V : Yt)(t, (function(t, r, o) { return e(n, t, r, o) })), n } }, 77: e => { "use strict";
                    e.exports = t }, 529: t => { "use strict";
                    t.exports = e } },
            r = {};

        function o(t) { var e = r[t]; if (void 0 !== e) return e.exports; var u = r[t] = { id: t, loaded: !1, exports: {} }; return n[t].call(u.exports, u, u.exports, o), u.loaded = !0, u.exports }
        o.n = t => { var e = t && t.__esModule ? () => t.default : () => t; return o.d(e, { a: e }), e }, o.d = (t, e) => { for (var n in e) o.o(e, n) && !o.o(t, n) && Object.defineProperty(t, n, { enumerable: !0, get: e[n] }) }, o.g = function() { if ("object" == typeof globalThis) return globalThis; try { return this || new Function("return this")() } catch (t) { if ("object" == typeof window) return window } }(), o.o = (t, e) => Object.prototype.hasOwnProperty.call(t, e), o.r = t => { "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(t, Symbol.toStringTag, { value: "Module" }), Object.defineProperty(t, "__esModule", { value: !0 }) }, o.nmd = t => (t.paths = [], t.children || (t.children = []), t); var u = {}; return (() => { "use strict";
            o.r(u), o.d(u, { clean: () => p, format: () => j, formatArray: () => d, formatBoolean: () => _, formatDate: () => h, formatNumber: () => v, formatObject: () => g, formatString: () => m, getType: () => c, trimArr: () => l, trimObj: () => s }); var t = o(826),
                e = o.n(t),
                n = o(77),
                r = o(923),
                i = o.n(r),
                a = o(53);

            function c(t) { return e = t, (0, a.containsThis)(e, a.hhmm) || (0, a.containsThis)(e, a.hhmmss) || (0, a.containsThis)(e, a.dashDDMMYYYY) || (0, a.containsThis)(e, a.dashYYYYMMDD) || (0, a.containsThis)(e, a.hyphenDDMMYYYY) || (0, a.containsThis)(e, a.hyphenYYYYMMDD) || (0, a.containsThis)(e, a.yyyymmdd) || (0, a.containsThis)(e, a.knexDate) || (0, a.containsThis)(e, a.stringDate) || e instanceof Date ? "date" : i().number(t) ? "number" : i().array(t) ? "array" : i().object(t) ? "object" : i().string(t) ? "string" : "boolean" == typeof t ? "boolean" : (0, n.upset)("not a valid type " + c, "getType in make legit", "should be a valid type"); var e }

            function f(t, e) {
                (null == e || e > t.length) && (e = t.length); for (var n = 0, r = new Array(e); n < e; n++) r[n] = t[n]; return r }

            function s(t) { if (t instanceof Date) return t;
                t = e()(t); for (var n = 0, r = Object.entries(t); n < r.length; n++) { var o = (c = r[n], s = 2, function(t) { if (Array.isArray(t)) return t }(c) || function(t, e) { var n = null == t ? null : "undefined" != typeof Symbol && t[Symbol.iterator] || t["@@iterator"]; if (null != n) { var r, o, u = [],
                                    i = !0,
                                    a = !1; try { for (n = n.call(t); !(i = (r = n.next()).done) && (u.push(r.value), !e || u.length !== e); i = !0); } catch (t) { a = !0, o = t } finally { try { i || null == n.return || n.return() } finally { if (a) throw o } } return u } }(c, s) || function(t, e) { if (t) { if ("string" == typeof t) return f(t, e); var n = Object.prototype.toString.call(t).slice(8, -1); return "Object" === n && t.constructor && (n = t.constructor.name), "Map" === n || "Set" === n ? Array.from(t) : "Arguments" === n || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n) ? f(t, e) : void 0 } }(c, s) || function() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.") }()),
                        u = o[0],
                        i = o[1]; if ("string" == typeof i) { var a = i.trim();
                        t[u] = a } } var c, s; return t }

            function l(t) { for (var e = [], n = 0; n < t.length; n++) { var r = c(t[n]); if ("string" === r) t[n] = t[n].trim(), e.push(t[n]);
                    else if ("object" === r) { var o = s(t[n]);
                        e.push(o) } else e.push(t[n]) } return e }

            function p(t) { switch (c(t)) {
                    case "array":
                        return l(t);
                    case "object":
                        return s(t);
                    case "string":
                        return t.trim();
                    default:
                        return t } } var y = o(529);

            function b(t) { return b = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(t) { return typeof t } : function(t) { return t && "function" == typeof Symbol && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t }, b(t) }

            function h(t) { var e = c(t); if ("string" === e) return (0, y.formatDateToPost)(t); if ("object" === e) { var n = (0, y.formatDateToPost)(t); return new Date(n) } }

            function v(t) { var e = c(t); if ("string" === e) return parseInt(t); if ("array" === e) { for (var r = 0, o = 0; o < t.length; o++) "number" == typeof t[o] && (r += t[o]); return r } if ("date" === e) { var u = (0, y.format)(t, "YYYYMMDDHHMMSS"); return parseInt(u) } return "number" == typeof t ? t : (0, n.upset)("not a valid type" + t, "format.js in make_legit", "should be a valid number") }

            function d(t) { var e = [],
                    n = c(t); if ("string" === n) e.push(t);
                else if ("number" === n) e.push(t);
                else if ("object" === n)
                    for (var r = "", o = Object.keys(t), u = Object.values(t), i = 0; i < o.length; i++) r += o[i].toString() + ": " + u[i].toString(), e.push(r);
                else if ("date" === n) { var a = (0, y.format)(t, "ll");
                    e.push(a) } else "boolean" === n ? e.push(t) : e = void 0; return Array.isArray(e) ? e : void 0 }

            function g(t) { var e = {}; if ("array" === c(t))
                    for (var n = 0; n < t.length; n++) e[n] = t[n];
                else e = void 0; return "object" === b(e) ? e : void 0 }

            function m(t) { var e, n = c(t); if ("number" === n) e = t.toString();
                else if ("array" === n) { for (var r = "", o = 0; o < t.length; o++) "object" === b(t[o]) ? r += JSON.stringify(t[o]) : r += t[o];
                    e = r } else e = "object" === n ? JSON.stringify(t) : "date" === n ? (0, y.format)(t, "ll") : "boolean" === n ? t.toString() : void 0; return "string" == typeof e ? e : void 0 }

            function _(t) { var e; if ("string" === c(t)) { var n = t.toLowerCase(); return "true" === n ? e = !0 : "false" === n && (e = !1), "boolean" == typeof e ? e : void 0 } }

            function j(t, e) { return c(t) === e ? p(t) : "date" === e ? h(t) : "string" === e ? m(t) : "array" === e ? p(d(t)) : "object" === e ? p(g(t)) : "boolean" === e ? _(t) : "number" === e ? v(t) : (0, n.upset)("not a valid type " + c, "getType in make legit", "should be a valid type") } })(), u })() }));
