// effective_knex/dist/counts.js
const { see } = require('code_clarity')
const { upset } = require("whats_wrong")
const knex = require("knex")({
    client: "postgresql",
    connection: {
        database: "optee_test",
        user: "postgres",
        password: "postgres",
    }
})



/**
 * @example
 *  let getCount = await tableColValCount(knex, table, col, val)
 * @author zen-out
 * @date 2022-03-10
 * @returns {number}
 */
function tableColValCount(knex, table, col, val) {
    let newObj = {}
    newObj[col] = val
    try {
        return knex(table).where(newObj).count().then((result) => {
            let finalCount = parseInt(result[0].count)
            return finalCount;
        })
    } catch (error) {
        upset(error, "counts.js", "should be able to get count")
    }
}
/**
 * @example
 * await getObjectCount(knex, table, object)
 * @author zen-out
 * @date 2022-03-23
 * @returns {number}
 */
async function getObjectCount(knex, table, object) {

    try {
        let count = await knex(table).where(object).count()
        let finalCount = parseInt(count[0].count)
        return finalCount;
    } catch (error) {
        upset(error, "counts.js", "should be able to get count")
    }
}
async function testCount() {
    // let getCount = await tableColValCount(knex, "problem", "user_id", 4)
    // if (typeof getCount === "number") {
    //     see.done("test passes")
    // } else {
    //     see.problem("test does not pass")
    // }
    let test_object_count = await getObjectCount(knex, "problem", { user_id: 100 })
    console.log(test_object_count)
}

testCount()



module.exports = { tableColValCount, getObjectCount, testCount };
