// effective_knex/module/basic.js
const _ = require("lodash")
const make_legit = require("make_legit")
const { makeFakes } = require("make_fakes")
const { getObject } = require("./getObject")
const { inputs, callbacks, } = require("../data")
const knex = require("knex")({
    client: "postgresql",
    connection: {
        database: "hbs_auth_temp",
        user: "postgres",
        password: "postgres",
    }
})

async function getOne(knex, table, id) {
    let object = await knex(table).select("*").where({ id: id })
    return getObject(object);
}

async function getAll(knex, table) {
    let all = await knex(table).select("*")
    return all
}
async function del(knex, table, id) {
    let deleted = await knex(table).where({ id: id }).del()

    return deleted;
}
module.exports = { getOne, getAll, del }
    // getOne(knex, "hourglass", 11).then((result) => {
    //     console.log(result)
    // })
    // getAll(knex, "hourglass").then((result) => {
    //     console.log(result)
    // })
    // del(knex, "hourglass", 54).then((result) => {
    //     console.log(result)
    // })


// module.exports = { getOne, getAll, del }
